File: dutch-server/src/services/BotDifficulty.ts
export interface BotDifficultyConfig {
  name: string;
  forgetChancePerTurn: number;
  confusionOnSwap: number;
  dutchThreshold: number;
  reactionSpeed: number;
  matchAccuracy: number;
  reactionMatchChance: number;
}

export class BotDifficulty {
  static readonly bronze: BotDifficultyConfig = {
    name: 'Bronze',
    forgetChancePerTurn: 0.18,
    confusionOnSwap: 0.3,
    dutchThreshold: 10,
    reactionSpeed: 0.55,
    matchAccuracy: 0.75,
    reactionMatchChance: 0.35,
  };

  static readonly silver: BotDifficultyConfig = {
    name: 'Argent',
    forgetChancePerTurn: 0.08,
    confusionOnSwap: 0.12,
    dutchThreshold: 6,
    reactionSpeed: 0.75,
    matchAccuracy: 0.85,
    reactionMatchChance: 0.55,
  };

  // Palier fort unique : fusion des anciens Or+Platine (= valeurs Platine, les
  // plus fortes), alignée sur le client (BotDifficulty.difficult, 'Difficile').
  static readonly difficult: BotDifficultyConfig = {
    name: 'Difficile',
    forgetChancePerTurn: 0,
    confusionOnSwap: 0,
    dutchThreshold: 1,
    reactionSpeed: 1,
    matchAccuracy: 1,
    reactionMatchChance: 1,
  };

  static fromMMR(mmr: number): BotDifficultyConfig {
    if (mmr < 300) {
      return this.bronze;
    } else if (mmr < 600) {
      return this.silver;
    } else {
      // Anciennes plages Or (600-899) + Platine (>=900) fusionnées.
      return this.difficult;
    }
  }

  static fromRank(rank: string): BotDifficultyConfig {
    switch (rank) {
      case 'Bronze':
        return this.bronze;
      case 'Argent':
        return this.silver;
      // Legacy 'Or'/'Platine' fusionnés en 'Difficile'.
      case 'Or':
      case 'Platine':
      case 'Difficile':
        return this.difficult;
      default:
        return this.silver;
    }
  }
}

File: lib/services/game/bot/bot_difficulty.dart
class BotDifficulty {
  final String name;
  final double forgetChancePerTurn;
  final double confusionOnSwap;
  final double reactionSpeed;
  final double matchAccuracy;
  final double reactionMatchChance;

  const BotDifficulty({
    required this.name,
    required this.forgetChancePerTurn,
    required this.confusionOnSwap,
    required this.reactionSpeed,
    required this.matchAccuracy,
    required this.reactionMatchChance,
  });

  static const BotDifficulty bronze = BotDifficulty(
    name: "Bronze",
    forgetChancePerTurn: 0.32,
    confusionOnSwap: 0.45,
    reactionSpeed: 0.35,
    matchAccuracy: 0.42,
    reactionMatchChance: 0.22,
  );

  static const BotDifficulty silver = BotDifficulty(
    name: "Argent",
    forgetChancePerTurn: 0.24,
    confusionOnSwap: 0.32,
    reactionSpeed: 0.50,
    matchAccuracy: 0.60,
    reactionMatchChance: 0.34,
  );

  /// Profil fort unique fusionné depuis les anciens paliers Or/Platine.
  /// La précision mécanique reste au niveau Platine; la stratégie métier
  /// filtre ensuite les coups trop agressifs.
  static const BotDifficulty difficult = BotDifficulty(
    name: "Difficile",
    forgetChancePerTurn: 0.0,
    confusionOnSwap: 0.0,
    reactionSpeed: 1.0,
    matchAccuracy: 1.0,
    reactionMatchChance: 1.0,
  );

  static const BotDifficulty gold = difficult;
  static const BotDifficulty platinum = difficult;

  static BotDifficulty fromMMR(int mmr) {
    if (mmr < 300) {
      return bronze;
    } else if (mmr < 600) {
      return silver;
    } else {
      return difficult;
    }
  }

  static BotDifficulty fromRank(String rank) {
    switch (rank) {
      case "Bronze":
        return bronze;
      case "Argent":
        return silver;
      case "Or":
      case "Platine":
      case "Difficile":
        return difficult;
      default:
        return silver;
    }
  }

  @override
  String toString() => name;
}
